File: parser/edi_parser.py
import re
from typing import Optional


def parse_edi(edi_text: str) -> dict:
    """Parse an EDI 837 claim file into a structured dict."""
    segments = [s.strip() for s in edi_text.replace("\n", "").split("~") if s.strip()]
    elements = [seg.split("*") for seg in segments]

    claim = {
        "patient": {
            "name": "",
            "member_id": "",
            "address": "",
            "city": "",
            "state": "",
            "zip": "",
            "dob": "",
            "gender": "",
        },
        "claim": {
            "id": "",
            "total_amount": 0.0,
            "place_of_service": "",
            "dates_of_service": "",
        },
        "provider": {
            "name": "",
            "npi": "",
            "address": "",
            "city": "",
            "state": "",
            "zip": "",
        },
        "billing_provider": {
            "name": "",
            "npi": "",
        },
        "plan": {
            "name": "",
            "payer_id": "",
        },
        "diagnoses": [],
        "procedures": [],
        "prior_auth": None,
        "raw_segment_count": len(segments),
    }

    i = 0
    current_loop = None
    address_buffer = {}

    while i < len(elements):
        seg = elements[i]
        seg_id = seg[0]

        if seg_id == "NM1":
            entity_code = seg[1] if len(seg) > 1 else ""
            name_last = seg[3] if len(seg) > 3 else ""
            name_first = seg[4] if len(seg) > 4 else ""
            name_mi = seg[5] if len(seg) > 5 else ""
            id_qualifier = seg[8] if len(seg) > 8 else ""
            id_value = seg[9] if len(seg) > 9 else ""

            if entity_code == "IL":  # Subscriber / patient
                full_name = f"{name_first} {name_last}".strip()
                claim["patient"]["name"] = full_name
                if id_qualifier == "MI":
                    claim["patient"]["member_id"] = id_value
                current_loop = "patient"
            elif entity_code == "85":  # Billing provider
                claim["billing_provider"]["name"] = name_last  # org name in element 3
                if id_qualifier == "XX":
                    claim["billing_provider"]["npi"] = id_value
                current_loop = "billing_provider"
            elif entity_code == "82":  # Rendering provider
                full_name = f"{name_first} {name_last}".strip()
                claim["provider"]["name"] = full_name
                if id_qualifier == "XX":
                    claim["provider"]["npi"] = id_value
                current_loop = "rendering_provider"
            elif entity_code == "40" or entity_code == "PR":  # Payer
                claim["plan"]["name"] = name_last
                claim["plan"]["payer_id"] = id_value
                current_loop = "payer"

        elif seg_id == "N3":
            address = seg[1] if len(seg) > 1 else ""
            address_buffer["street"] = address

        elif seg_id == "N4":
            city = seg[1] if len(seg) > 1 else ""
            state = seg[2] if len(seg) > 2 else ""
            zip_code = seg[3] if len(seg) > 3 else ""
            address_buffer.update({"city": city, "state": state, "zip": zip_code})

            street = address_buffer.get("street", "")
            full_address = f"{street}, {city}, {state} {zip_code}".strip(", ")

            if current_loop == "patient":
                claim["patient"]["address"] = street
                claim["patient"]["city"] = city
                claim["patient"]["state"] = state
                claim["patient"]["zip"] = zip_code
            elif current_loop == "billing_provider":
                claim["billing_provider"]["address"] = full_address
            elif current_loop == "rendering_provider":
                claim["provider"]["address"] = full_address

            address_buffer = {}

        elif seg_id == "DMG":
            dob = seg[2] if len(seg) > 2 else ""
            gender = seg[3] if len(seg) > 3 else ""
            if len(dob) == 8:
                dob = f"{dob[:4]}-{dob[4:6]}-{dob[6:]}"
            claim["patient"]["dob"] = dob
            claim["patient"]["gender"] = gender

        elif seg_id == "CLM":
            claim["claim"]["id"] = seg[1] if len(seg) > 1 else ""
            try:
                claim["claim"]["total_amount"] = float(seg[2]) if len(seg) > 2 else 0.0
            except ValueError:
                claim["claim"]["total_amount"] = 0.0
            claim["claim"]["place_of_service"] = _parse_pos(seg[5]) if len(seg) > 5 else ""

        elif seg_id == "DTP":
            qualifier = seg[1] if len(seg) > 1 else ""
            if qualifier == "434":
                claim["claim"]["dates_of_service"] = seg[3] if len(seg) > 3 else ""

        elif seg_id == "REF":
            qualifier = seg[1] if len(seg) > 1 else ""
            value = seg[2] if len(seg) > 2 else ""
            if qualifier == "9F":  # Prior authorization number
                claim["prior_auth"] = value

        elif seg_id == "HI":
            for elem in seg[1:]:
                if elem:
                    parts = elem.split(":")
                    code = parts[1] if len(parts) > 1 else parts[0]
                    if code:
                        claim["diagnoses"].append(code.strip())

        elif seg_id == "SBR":
            plan_name = seg[3] if len(seg) > 3 else ""
            if plan_name:
                claim["plan"]["name"] = plan_name

        elif seg_id == "LX":
            current_loop = "service_line"

        elif seg_id == "SV1":
            proc_raw = seg[1] if len(seg) > 1 else ""
            charge = seg[3] if len(seg) > 3 else "0"
            units = seg[5] if len(seg) > 5 else "1"

            parts = proc_raw.split(":")
            qualifier = parts[0] if parts else ""
            proc_code = parts[1] if len(parts) > 1 else proc_raw

            try:
                charge_val = float(charge)
            except ValueError:
                charge_val = 0.0

            try:
                units_val = float(units)
            except ValueError:
                units_val = 1.0

            claim["procedures"].append({
                "code": proc_code.strip(),
                "qualifier": qualifier.strip(),
                "charge": charge_val,
                "units": units_val,
            })

        i += 1

    claim["diagnoses"] = list(dict.fromkeys(claim["diagnoses"]))
    return claim


def _parse_pos(pos_composite: str) -> str:
    """Extract place of service from composite element like '11:B:1'."""
    parts = pos_composite.split(":")
    pos_map = {
        "11": "Office",
        "21": "Inpatient Hospital",
        "22": "Outpatient Hospital",
        "23": "Emergency Room",
    }
    return pos_map.get(parts[0], parts[0]) if parts else pos_composite


def summarise_claim(parsed: dict) -> str:
    """Return a one-line human-readable summary of a parsed claim."""
    procs = ", ".join(p["code"] for p in parsed["procedures"])
    diags = ", ".join(parsed["diagnoses"])
    amount = parsed["claim"]["total_amount"]
    return (
        f"Claim {parsed['claim']['id']} | "
        f"Amount: ${amount:,.2f} | "
        f"Procedures: {procs} | "
        f"Diagnoses: {diags} | "
        f"Auth: {parsed['prior_auth'] or 'NONE'}"
    )

File: frontend/src/lib/ediParser.ts
const POS_MAP: Record<string, string> = {
  "11": "Office",
  "21": "Inpatient Hospital",
  "22": "Outpatient Hospital",
  "23": "Emergency Room",
};

export interface Procedure {
  code: string;
  qualifier: string;
  charge: number;
  units: number;
}

export interface ParsedClaim {
  patient: {
    name: string;
    member_id: string;
    address: string;
    city: string;
    state: string;
    zip: string;
    dob: string;
    gender: string;
  };
  claim: {
    id: string;
    total_amount: number;
    place_of_service: string;
    dates_of_service: string;
  };
  provider: {
    name: string;
    npi: string;
    address: string;
    city: string;
    state: string;
    zip: string;
  };
  billing_provider: {
    name: string;
    npi: string;
  };
  plan: {
    name: string;
    payer_id: string;
  };
  diagnoses: string[];
  procedures: Procedure[];
  prior_auth: string | null;
  raw_segment_count: number;
}

export function parseEdi(ediText: string): ParsedClaim {
  const segments = ediText
    .replace(/\n/g, "")
    .split("~")
    .map((s) => s.trim())
    .filter(Boolean);
  const elements = segments.map((seg) => seg.split("*"));

  const claim: ParsedClaim = {
    patient: { name: "", member_id: "", address: "", city: "", state: "", zip: "", dob: "", gender: "" },
    claim: { id: "", total_amount: 0, place_of_service: "", dates_of_service: "" },
    provider: { name: "", npi: "", address: "", city: "", state: "", zip: "" },
    billing_provider: { name: "", npi: "" },
    plan: { name: "", payer_id: "" },
    diagnoses: [],
    procedures: [],
    prior_auth: null,
    raw_segment_count: segments.length,
  };

  let currentLoop: string | null = null;
  let addressBuffer: Record<string, string> = {};

  for (const seg of elements) {
    const segId = seg[0];

    if (segId === "NM1") {
      const entityCode = seg[1] || "";
      const nameLast = seg[3] || "";
      const nameFirst = seg[4] || "";
      const idQualifier = seg[8] || "";
      const idValue = seg[9] || "";

      if (entityCode === "IL") {
        claim.patient.name = `${nameFirst} ${nameLast}`.trim();
        if (idQualifier === "MI") claim.patient.member_id = idValue;
        currentLoop = "patient";
      } else if (entityCode === "85") {
        claim.billing_provider.name = nameLast;
        if (idQualifier === "XX") claim.billing_provider.npi = idValue;
        currentLoop = "billing_provider";
      } else if (entityCode === "82") {
        claim.provider.name = `${nameFirst} ${nameLast}`.trim();
        if (idQualifier === "XX") claim.provider.npi = idValue;
        currentLoop = "rendering_provider";
      } else if (entityCode === "40" || entityCode === "PR") {
        claim.plan.name = nameLast;
        claim.plan.payer_id = idValue;
        currentLoop = "payer";
      }
    } else if (segId === "N3") {
      addressBuffer.street = seg[1] || "";
    } else if (segId === "N4") {
      const city = seg[1] || "";
      const state = seg[2] || "";
      const zipCode = seg[3] || "";
      addressBuffer = { ...addressBuffer, city, state, zip: zipCode };

      const street = addressBuffer.street || "";
      if (currentLoop === "patient") {
        claim.patient.address = street;
        claim.patient.city = city;
        claim.patient.state = state;
        claim.patient.zip = zipCode;
      } else if (currentLoop === "billing_provider") {
        claim.billing_provider.npi = claim.billing_provider.npi;
      } else if (currentLoop === "rendering_provider") {
        claim.provider.address = `${street}, ${city}, ${state} ${zipCode}`.trim();
      }
      addressBuffer = {};
    } else if (segId === "DMG") {
      let dob = seg[2] || "";
      const gender = seg[3] || "";
      if (dob.length === 8) {
        dob = `${dob.slice(0, 4)}-${dob.slice(4, 6)}-${dob.slice(6)}`;
      }
      claim.patient.dob = dob;
      claim.patient.gender = gender;
    } else if (segId === "CLM") {
      claim.claim.id = seg[1] || "";
      claim.claim.total_amount = parseFloat(seg[2]) || 0;
      const posComposite = seg[5] || "";
      const posPart = posComposite.split(":")[0];
      claim.claim.place_of_service = POS_MAP[posPart] || posPart;
    } else if (segId === "DTP") {
      const qualifier = seg[1] || "";
      if (qualifier === "434") {
        claim.claim.dates_of_service = seg[3] || "";
      }
    } else if (segId === "REF") {
      const qualifier = seg[1] || "";
      const value = seg[2] || "";
      if (qualifier === "9F") {
        claim.prior_auth = value;
      }
    } else if (segId === "HI") {
      for (const elem of seg.slice(1)) {
        if (elem) {
          const parts = elem.split(":");
          const code = parts[1] || parts[0];
          if (code) claim.diagnoses.push(code.trim());
        }
      }
    } else if (segId === "LX") {
      currentLoop = "service_line";
    } else if (segId === "SV1") {
      const procRaw = seg[1] || "";
      const charge = seg[3] || "0";
      const units = seg[5] || "1";
      const parts = procRaw.split(":");
      const qualifier = parts[0] || "";
      const procCode = parts[1] || procRaw;

      claim.procedures.push({
        code: procCode.trim(),
        qualifier: qualifier.trim(),
        charge: parseFloat(charge) || 0,
        units: parseFloat(units) || 1,
      });
    }
  }

  claim.diagnoses = [...new Set(claim.diagnoses)];
  return claim;
}

const PHI_FIELDS = new Set(["name", "member_id", "address", "city", "zip", "dob"]);

function makeToken(field: string): string {
  const uid = Math.random().toString(16).slice(2, 10).toUpperCase();
  if (field === "name") return `FNAME-${uid}`;
  if (field === "member_id") return `MID-${uid}`;
  if (["address", "city", "zip"].includes(field)) return `ADDR-${uid}`;
  if (field === "dob") return `DOB-${uid}`;
  return `PHI-${uid}`;
}

export function maskPhi(claim: ParsedClaim): { masked: Record<string, unknown>; tokenMap: Record<string, string> } {
  const tokenMap: Record<string, string> = {};

  function maskValue(value: unknown, fieldName: string | null): unknown {
    if (value === null || value === undefined) return value;
    if (typeof value === "object" && !Array.isArray(value)) {
      const result: Record<string, unknown> = {};
      for (const [k, v] of Object.entries(value as Record<string, unknown>)) {
        result[k] = maskValue(v, k);
      }
      return result;
    }
    if (Array.isArray(value)) {
      return value.map((item) => maskValue(item, fieldName));
    }
    if (typeof value === "string" && value.trim()) {
      if (fieldName && PHI_FIELDS.has(fieldName)) {
        const existing = Object.entries(tokenMap).find(([, v]) => v === value);
        if (existing) return existing[0];
        const token = makeToken(fieldName);
        tokenMap[token] = value;
        return token;
      }
    }
    return value;
  }

  const masked = maskValue(claim, null) as Record<string, unknown>;
  return { masked, tokenMap };
}
